src/main/scala/scala/compat/java8/StreamConverters.scala

```diff
@@ -0,0 +1,339 @@
+  extends MakesSequentialStream[java.lang.Long, LongStream] with MakesParallelStream[java.lang.Long, LongStream] {
+    def seqStream: LongStream = StreamSupport.longStream(steppize(cc).stepper, false)
+    def parStream: LongStream = StreamSupport.longStream(steppize(cc).stepper.anticipateParallelism, true)
+  }
+  implicit class EnrichLongKeySteppableWithStream[CC](cc: CC)(implicit steppize: CC => MakesLongKeyStepper) {
+    def seqKeyStream: LongStream = StreamSupport.longStream(steppize(cc).keyStepper, false)
+    def parKeyStream: LongStream = StreamSupport.longStream(steppize(cc).keyStepper.anticipateParallelism, true)
+  }
+  implicit class EnrichLongValueSteppableWithStream[CC](cc: CC)(implicit steppize: CC => MakesLongValueStepper) {
+    def seqValueStream: LongStream = StreamSupport.longStream(steppize(cc).valueStepper, false)
+    def parValueStream: LongStream = StreamSupport.longStream(steppize(cc).valueStepper.anticipateParallelism, true)
+  }
+}
+
+trait Priority1StreamConverters extends Priority2StreamConverters {
+  implicit class RichStream[A](stream: Stream[A]) {
+    def accumulate = stream.collect(Accumulator.supplier[A], Accumulator.adder[A], Accumulator.merger[A])
+    
+    def toScala[Coll[_]](implicit cbf: collection.generic.CanBuildFrom[Nothing, A, Coll[A]]): Coll[A] = {
+      if (stream.isParallel) accumulate.to[Coll](cbf)
+      else {
+        val b = cbf()
+        stream.forEachOrdered(new java.util.function.Consumer[A]{ def accept(a: A) { b += a } })
+        b.result()
+      }
+    }
+    
+    def unboxed[S](implicit ubx: PrimitiveStreamUnboxer[A, S]): S = ubx(stream)
+  }
+  
+  implicit class RichStreamCanAccumulatePrimitive[S](stream: S) {
+    def accumulatePrimitive[AA](implicit psa: PrimitiveStreamAccumulator[S, AA]) = psa.streamAccumulate(stream)
+  }
+}
+
+/** `StreamConverters` provides extension methods and other functionality to
+  * ease interoperability of Scala collections with `java.util.stream` classes.
+  * 
+  * Scala collections gain extension methods `seqStream` and
+  * `parStream` that allow them to be used as the source of a `Stream`.
+  * Some collections either intrinsically cannot be paralellized, or
+  * could be but an efficient implementation is missing.  It this case,
+  * only `seqStream` is provided.  If a collection cannot be stepped over
+  * at all (e.g. `Traversable`), then it gains neither method.
+  *
+  * `Array` also gains `seqStream` and `parStream` methods, and calling those
+  * on `Array[Double]`, `Array[Int]`, or `Array[Long]` will produce the
+  * corresponding primitive stream.
+  *
+  * Streams gain `accumulate` and `toScala[_]` methods, which collect the stream
+  * into a custom high-performance `scala.collection.mutable.java8.Accumulator`,
+  * which is not part of the standard collections hierarchy, or into a named
+  * Scala collection, respectively.
+  *
+  * Generic streams also gain an `unboxed` method that will convert to the
```

Comment:
> But what if you want a stream of Char? There is no primitive type that corresponds. Do we want to make this intentionally awkward? (What do we even do?)

The supported primitive stream types can support all other primitive types through widening conversions, so I'd say use `IntStream` for `Char`. To make this feasible, we'd also need a way to perform a narrowing conversion when collecting into an unboxed primitive collection (like `Array[Char]`).

> I already have designed the priority hierarchy such that if you have a collection of Double you can't conveniently get a Stream instead of a DoubleStream out of it. (I hope. That was my goal.)

The loophole is abstracting over types. The current design allows you to get a Stream for a collection of `T <: Any` and since it's implemented through an extension method with static dispatch, all primitive types are boxed. Restricting the supported types to `T <: AnyRef` would make that illegal, so you're forced to special-case primitive types (or cast if you're willing to accept the boxing overhead).